compiler/rustc_mir_build/src/build/mod.rs

```diff
@@ -71,6 +72,16 @@ fn mir_build(tcx: TyCtxt<'_>, def: ty::WithOptConstParam<LocalDefId>) -> Body<'_
             // of `mir_build`, so now we can steal it
             let thir = thir.steal();
 
+            if def.const_param_did.is_none() {
+                parallel!({ tcx.ensure().check_match(def.did) }, {
```

Comment:
I would avoid parallelism deeper in the query system to avoid the deadlock risk that brings.